### PR TITLE
Compat blog post take 2

### PR DESCRIPTION
**Pull request by Nemo157**
Replacement for #1226 (@MajorBreakfast I hope you're ok with me taking this over).

I still want to add some info on the spawn and IO compat that's been added since this was originally written, and I need to actually test the snippets against the new API. Should get to that over the weekend at some point.

### CODE REVIEW

File: _posts/2018-01-01-compatibility-layer.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+[`futures@0.1`]: https://docs.rs/futures
+[`futures@0.3`]: https://rust-lang-nursery.github.io/futures-api-docs/
+[`std::future`]: https://doc.rust-lang.org/nightly/std/future/
+
+## `Cargo.toml`
+
+The compatibility layer can be enabled by setting the `compat` feature in your `Cargo.toml`:
+
+```toml
+futures-preview = { version = "0.3.0-alpha.10", features = ["compat"] }
```
Comment (cramertj) [inline _posts/2018-01-01-compatibility-layer.md]: `compat` doesn't include `io-compat`, which also pulls in `tokio-io` and adds the `AsyncRead` and `AsyncWrite` compat bits.

**Comment by cramertj** [comment]
Aside from the one nit and updating to add IO/spawn info, this LGTM so far!

**Comment by LucioFranco** [comment]
One, thing that might be useful to add is something about sink support with tokio. As I used this post to convert my project to async/await. I was a bit confused on how to approach sinks. I ended up with using some of the tokio compat features. So maybe a mention about it would be helpful imo.

File: _posts/2018-01-01-compatibility-layer.md (hunk 1 above)
Comment (Nemo157) [reply _posts/2018-01-01-compatibility-layer.md]: Yep, I was going to mention the extra features when adding the additional IO/spawn sections.

**Comment by Nemo157** [comment]
@LucioFranco yeah, probably worth adding a section about `Sink`. Currently it's possible to wrap a 0.3 `Sink` into a 0.1 `Sink`, but not the other direction. I will also open an issue about whether it's possible to add that additional conversion, I couldn't think of an easy way when implementing the current `Sink` compat stuff but maybe it's possible with a little bit of extra runtime cost.

**Comment by Nemo157** [comment]
Added info on `Spawn`, IO and `Sink`, and updated the final example to use the new IO compat layer.

**Comment by LucioFranco** [comment]
@Nemo157 Great! I just took a look at it again, it def clears some things up. I am still slightly struggling to understand where the `TokioDefaultSpawner` fits into the whole picture. Also, I think the fact that there is no `0.1` -> `0.3` is a bigger issue for people using tokio. So maybe might be worthwhile to be more upfront about that?

This btw is my hacky solution to using std await and `UdpFramed` from tokio. I would be interested to help build out some of the sink compat stuff.
```rust
while let Some(msg) = await!(rx.next()) {
    println!("Sending: {:?}", msg);

    {
        use tokio::prelude::SinkExt;
        if let Err(e) = await!(sink.send_async((msg, to))) {
            println!("Sending message: {}", e);
        }
    }
}
```

I would also maybe add a `cargo.toml` snippet for the example so that its easier for people to try it out.

**Comment by Nemo157** [comment]
@LucioFranco thanks, those are some good points. I think an example of using `TokioDefaultSpawner` is probably the easiest way to show what it’s useful for. I can expand a bit more on the `Sink` situation as well. And having the `Cargo.toml` for the example would definitely be worth it.

**Comment by Nemo157** [comment]
Sorry about the big delay on this again, I'll try and finish this off before 2019!

**Comment by Nemo157** [comment]
Updated now that `Sink` is supported and `TokioDefaultSpawner` was removed, also uses the new methods from #1469 for consistency so is blocked on that being merged and a new release.

**Comment by corbinu** [comment]
@Nemo157 Read through it as somebody outside of the working team and just excited about async/await. It looks great and is easy to understand. Super excited!

**Comment by archseer** [comment]
How would one go about using `select!` with tokio compatibility? I ran into some problems because `select-macro` seems to call out to the stdlib/futures implementations so it doesn't hit the [`tokio` await wrapper](https://github.com/tokio-rs/tokio/blob/9c5cad037fd78adb087bc4734f7670d5732122c5/tokio-async-await/src/await.rs).

**Comment by Nemo157** [comment]
@archseer I have no real experience using Tokio's `await!` macro, but you can just use the `futures` 0.1 -> 0.3 compat methods to convert Tokio's futures into `std` futures so that they work with `std::await!`, something like

```rust
use futures::future::FutureExt;
use futures::compat::Future01CompatExt;

select! {
        foo = some_tokio_future().compat().fuse() => { ... },
        bar = some_other_tokio_future().compat().fuse() => { ... },
};
```

**Comment by Nemo157** [comment]
@cramertj I've checked the example code works with the new release and all the links are correct, I think this is good to publish finally 😁

**Comment by cramertj** [comment]
Nice!

**Pull request merged**